core/modules/replica/Request.cc

```diff
@@ -0,0 +1,270 @@
+        _workerInfo(serviceProvider->config()->workerInfo(worker)),
+        _timerIvalSec(serviceProvider->config()->retryTimeoutSec()),
+        _timer(io_service),
+        _requestExpirationIvalSec(serviceProvider->config()->controllerRequestTimeoutSec()),
+        _requestExpirationTimer(io_service) {
+
+        _serviceProvider->assertWorkerIsValid(worker);
+}
+std::string Request::context() const {
+    return "REQUEST " + id() + "  " + type() +
+           "  " + state2string(state(), extendedState()) +
+           "::" + replica::status2string(extendedServerStatus()) + "  ";
+}
+
+std::string const& Request::remoteId() const {
+    return _duplicateRequestId.empty() ? _id : _duplicateRequestId;
+}
+
+void Request::start(std::shared_ptr<Controller> const& controller,
+                    std::string const& jobId,
+                    unsigned int requestExpirationIvalSec) {
+
+    LOCK_GUARD;
+
+    assertState(CREATED);
+
+    // Change the expiration ival if requested
+    if (requestExpirationIvalSec) {
+        _requestExpirationIvalSec = requestExpirationIvalSec;
+    }
+    LOGS(_log, LOG_LVL_DEBUG, context() << "start  _requestExpirationIvalSec: "
+         << _requestExpirationIvalSec);
+
+    // Build optional associaitons with the corresponding Controller and the job
+    //
+    // NOTE: this is done only once, the first time a non-trivial value
+    // of each parameter is presented to the method.
+
+    if (not _controller    and     controller)    { _controller = controller; }
+    if (    _jobId.empty() and not jobId.empty()) { _jobId      = jobId; }
+
+    _performance.setUpdateStart();
+
+    if (_requestExpirationIvalSec) {
+        _requestExpirationTimer.cancel();
+        _requestExpirationTimer.expires_from_now(boost::posix_time::seconds(_requestExpirationIvalSec));
+        _requestExpirationTimer.async_wait(
+            boost::bind(
+                &Request::expired,
+                shared_from_this(),
+                boost::asio::placeholders::error
+            )
+        );
+    }
+
+    // Let a subclass to proceed with its own sequence of actions
+    startImpl();
+
+    _controller->serviceProvider()->databaseServices()->saveState(shared_from_this());
+}
+
+std::string const& Request::jobId() const {
+    if (_state == State::CREATED) {
+        throw std::logic_error(
+            "the Job Id is not available because the request has not started yet");
+    }
+    return _jobId;
+}
+
+void Request::expired(boost::system::error_code const& ec) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "expired"
+         << (ec == boost::asio::error::operation_aborted ? "  ** ABORTED **" : ""));
+
+    // Ignore this event if the timer was aborted
+    if (ec == boost::asio::error::operation_aborted) { return; }
+
+    // Also ignore this event if the request is over
+    if (_state == State::FINISHED) { return; }
+    {
+        LOCK_GUARD;
+        finish(EXPIRED);
+    }

```

Comment:
ok, this is just mean to whoever hast to come read this code ;-D - it looks like the block starting on line 190 is related to the if statement, but it's actually a context for the lock guard.
If it were me I would 

    if (_state == State::FINISHED) { 
        return; 
    }

    { // context for lock
        LOCK_GUARD;
        finish(EXPIRED);
    }